ENH Add samesite attribute to cookies.

GuySartorelli opened the pull request:
Sets a default `samesite` attribute on all cookies. Using the `Lax` value means that we're just setting an explicit value that is identical to the implicit value that was already being used.

This PR provides a way to configure both the default value for all cookies and for session cookies separately, as well as giving an extension hook if a specific value is desired for a given cookie.

The extension hook should be removed in the next major in favour of adding a new parameter like what was proposed in #9842 - I've avoided doing that here because the cookie backend is explicitly extensible (with an interface and docs and everything) so adding new parameters here would constitute a breaking change.

The following links may be useful background:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite
https://www.php.net/manual/en/function.setcookie.php
https://www.php.net/manual/en/function.session-set-cookie-params.php

## Parent issue:
- https://github.com/silverstripe/silverstripe-framework/issues/9440

## Related issue:
- tractorcow-farm/silverstripe-fluent#651

GuySartorelli commented:
Leaving in draft for now because I still need to add unit tests.

GuySartorelli commented:
Actually.... I don't think there _is_ a way to test this... none of the unit tests current test against any of the properties of cookies other than the values and whether they're set at all. I don't think there's a way we can test if the cookie's `samesite` attribute is set correctly.

Marking as ready for review. But if anyone has ideas for how we could test this, that'd be awesome.

GuySartorelli commented:
After giving this a go it looks like I can't test cookies in a `FunctionalTest` - the cookies are set using [`setcookie()`](https://www.php.net/manual/en/function.setcookie.php) directly, rather than adding them to the `HTTPResponse` object, which means we aren't fully including them in mock requests. The cookie _values_ are being set in mock requests by [pulling them out of the cookie jar](https://github.com/silverstripe/silverstripe-framework/blob/8ae7f30c417729292ba68cd0c500e02cc5106550/src/Control/Director.php#L230) and [setting the `$_COOKIE` env variable](https://github.com/silverstripe/silverstripe-framework/blob/8ae7f30c417729292ba68cd0c500e02cc5106550/src/Control/Director.php#L275). So we are mocking cookie _values_ but not the other attributes in cookie headers.

Changing the way we add cookies to make them more fully mockable would definitely be a valuable enhancement but is out of scope for this PR. I've created an issue (#10338) for that.

emteknetnz commented on docs/en/02_Developer_Guides/18_Cookies_And_Sessions/01_Cookies.md:
```diff
@@ -52,6 +56,44 @@ Cookie::force_expiry($name, $path = null, $domain = null);
+
+class CookieJarExtension extends Extension
+{
+    public function updateSameSite(string $cookieName, string $sameSite): void
```
```suggestion
    public function updateSameSite(string &$cookieName, string &$sameSite): void
```

emteknetnz commented on src/Control/CookieJar.php:
```diff
@@ -168,9 +170,17 @@ protected function outputCookie(
+                'expires' => $expiry ?? 0,
+                'path' => $path ?? '',
+                'domain' => $domain ?? '',
+                'secure' => ($sameSite === 'None') ? true : (bool)$secure,
```
```suggestion
                'secure' => $sameSite === 'None' ? true : (bool) $secure,
```

emteknetnz commented on docs/en/02_Developer_Guides/09_Security/04_Secure_Coding.md:
````diff
@@ -762,11 +762,15 @@ disable this behaviour using `CanonicalURLMiddleware::singleton()->setForceBasic
 
 ```yml
 SilverStripe\Control\Session:
+  cookies_samesite: 'Strict'
````
```suggestion
  cookie_samesite: 'Strict'
```
Remove the plural to match `cookie_secure`.  You'll need to update this in a bunch of places

emteknetnz commented on src/Control/Session.php:
```diff
@@ -135,6 +135,13 @@ class Session
 
+    /**
+     * Must be "Strict", "Lax", "None", or empty.
+     * If empty, the default value will be applied.
```
```suggestion
     * If blank string, the default value from `Cookie::config()->get('default_samesite')` will be applied.
```

emteknetnz commented on src/Control/Session.php:
```diff
@@ -135,6 +135,13 @@ class Session
     private static $cookie_name_secure = 'SECSESSID';
 
+    /**
+     * Must be "Strict", "Lax", "None", or empty.
```
```suggestion
     * Must be "Strict", "Lax", "None", or a blank string.
```

emteknetnz commented on src/Control/Session.php:
```diff
@@ -135,6 +135,13 @@ class Session
+     * If empty, the default value will be applied.
+     * @config
+     */
+    private static string $cookies_samesite = '';
```
I'm kind of inclined to just default set this to 'Lax' here and get rid of the fallback to Cookie::config()->get('default_samesite').  Seems less complex and confusing

GuySartorelli replied:
I'd find it rather confusing as a developer if I changed the default for all cookies to be `Strict` but the session cookie was coming through as `Lax` still. It's a global default for _all_ cookies.

emteknetnz replied:
> I'd find it rather confusing as a developer if I changed the default for all cookies to be Strict

The default in Cookie.default_samesite = 'Lax' though?

Or do you mean if we were to change this to 'Lax' and a developer left this as 'Lax' but changed Cookie.default_samesite = "Strict" then that would be confusing?  But, even if we left things as is, then a developer can do that anyway since we provide 2 config values

All the confusion surronding this is not a good sign

Is there actually a use case for all thing configurability?  Right now we seem to have:

- Cookie.default_samesite = "Lax"
- Session.cookie_samesite = ""
- CookieJarExtension::updateSameSite extension hook to update the SameSite attribute for individual cookie names

It really feels like there's too much here

GuySartorelli replied:
Yeah I can see that. I'll remove the fallback on `Cookie.default_samesite` for session cookies and update the documentation accordingly.

emteknetnz replied:
Also possibly better to just get rid of Session.cookie_samesite, rename Cookie.default_same to just Cookie.cookie_samesite.  Do we need 2 different config values here?

GuySartorelli replied:
> Get rid of Session.cookie_samesite

I don't know enough about how cookies are used to really make a call here - but my instinct is that there might be a usecase for "Lax" cookies generally - whereas for security purposes realistically the session cookie should be "Strict". We can't set it as strict by default because that could be a breaking change, but I think giving the ability to configure this separately is useful.

> rename Cookie.default_same to Cookie.cookie_samesite

I'd prefer to keep this name as-is. Even in a major release when we add a `samesite` parameter, I think it would be beneficial to retain this global configuration option (to configure the `samesite` attribute used for cookies in core or in modules), and the "default" especially at that time is important (since it will then be overridable by passing a different value to the parameter).

emteknetnz commented on src/Control/CookieJar.php:
```diff
@@ -180,4 +190,20 @@ protected function outputCookie(
+    private function getSameSite(string $name): string
+    {
+        if ($name === session_name()) {
+            return Session::config()->get('cookies_samesite') ?: Cookie::config()->get('default_samesite');
```
This logic is pretty confusing, I've made a comment later on about consolidating this config

emteknetnz commented on docs/en/02_Developer_Guides/18_Cookies_And_Sessions/01_Cookies.md:
````diff
@@ -52,6 +56,44 @@ Cookie::force_expiry($name, $path = null, $domain = null);
+```php
+<?php
+
+use SilverStripe\Core\Extension;
````
Why are we using an extension here instead of just adding an extra param? i.e. why would we need to wait for a new CMS version?  Seems bad to say "this is how you do the thing" and immediately deprecate it.

GuySartorelli replied:
Adding an extra parameter will break any custom classes that implement `Cookie_Backend`.

emteknetnz replied:
OK, makes sense that we're constrained by the interface

I'm inclined to just get rid of the extension hook functionality altogether and just not allow overriding on individual cookie names. I don't like the idea of adding functionality only to deprecate it straight away.  It just makes upgrading more painful.

Currently people are already using 'Lax' due to it being a browser default, so by not providing this functionality we are not going to break any implementations.

GuySartorelli replied:
That's fair. In a major release the deprecation notice there should prompt us to add the parameters, so people can set different `samesite` attributes to their specific cookies that way. In the meantime being able to change the value globally is still more control than people currently have so I'm okay with not providing the extension.
If someone does really want this then hopefully they'll create an issue or a PR and further discussion can be had at that time.

emteknetnz commented on src/Control/CookieJar.php:
```diff
@@ -180,4 +187,17 @@ protected function outputCookie(
+    /**
+     * Get the correct samesite value - Session cookies use a different configuration variable.
+     *
+     * @deprecated 5.0 The relevant methods will include a `$sameSite` parameter instead.
```
Deprecating this method is weird now the extension hook has been removed

Is there someway to 'deprecate' method signature on Cookie_Backed::set() to remind ourselves to add the extra parameter?

GuySartorelli replied:
We won't need this method once the attributes are added, because `Session::start()` will just pass the samesite value directly into its call to `Cookie::set()`.

emteknetnz commented on src/Control/Session.php:
```diff
@@ -307,7 +312,16 @@ public function start(HTTPRequest $request)
             if (!headers_sent()) {
-                session_set_cookie_params($timeout ?: 0, $path, $domain ?: null, $secure, true);
+                $sameSite = static::config()->get('cookie_samesite');
+                $secure = ($sameSite === 'None') ? true : Director::is_https($request) && $this->config()->get('cookie_secure');
```
Bit too much going on in this line, could you extra to a private method

emteknetnz commented on docs/en/02_Developer_Guides/18_Cookies_And_Sessions/02_Sessions.md:
```diff
@@ -101,7 +101,19 @@ including form and page comment information. None of this is vital but `clear_al
+### Samesite attribute
+
+The session cookie is handled slightly differently than most cookies on the site, which provides the opportunity to handle the samesite attribute separately from other cookies.
+You can also change the `samesite` attribute for session cookies like so:
```
```suggestion
You can change the `samesite` attribute for session cookies like so:
```

GuySartorelli commented:
@emteknetnz Requested changes made.

emteknetnz commented on docs/en/02_Developer_Guides/09_Security/04_Secure_Coding.md:
````diff
@@ -784,6 +788,12 @@ SilverStripe\Core\Injector\Injector:
 ```
 
+[info]
+There is not currently an easy way to pass a `securesite` attribute value for setting this cookie - but you can set the
````
Remove section

GuySartorelli replied:
I'll remove the bit that talks about adding an extension - but I think it's worth pointing at the `Cookie.default_samesite` config here.

emteknetnz commented on docs/en/02_Developer_Guides/09_Security/04_Secure_Coding.md:
```diff
@@ -762,11 +762,15 @@ disable this behaviour using `CanonicalURLMiddleware::singleton()->setForceBasic
 We also want to ensure cookies are not shared between secure and non-secure sessions, so we must tell Silverstripe CMS to 
-use a [secure session](https://docs.silverstripe.org/en/3/developer_guides/cookies_and_sessions/sessions/#secure-session-cookie). 
-To do this, you may set the `cookie_secure` parameter to `true` in your `config.yml` for `Session`
+use a [secure session](https://docs.silverstripe.org/en/3/developer_guides/cookies_and_sessions/sessions/#secure-session-cookie).
```
```suggestion
use a [secure session](https://docs.silverstripe.org/en/4/developer_guides/cookies_and_sessions/sessions/#secure-session-cookie).
```

Probably also seeing if there are other en/3/ links around, if so please open a new issue with corresponding PR to find and replace them

GuySartorelli replied:
I'll swap this to be a relative URL instead. This probably was carried over when the `4` branch was created and we want to avoid having to change it again in the next major release docs.

GuySartorelli replied:
I have checked and the only other link to v3 docs is intentional - but I've created an issue to swap all the absolute URLs to relative URLs so we don't have this issue in future majors.

emteknetnz commented on docs/en/04_Changelogs/4.12.0.md:
```diff
@@ -0,0 +1,55 @@
+
+The `samesite` attribute is now set on all cookies. To avoid backward compatability issues, the `Lax` value has been set by default, but we recommend reviewing the requirements of your project and setting an appropriate value.
+
+The default value can be set for all cookies in yaml configuration like so:
```
```suggestion
The default value can be set for all cookies (except for the Session cookie) in yaml configuration like so:
```

emteknetnz commented on docs/en/02_Developer_Guides/18_Cookies_And_Sessions/01_Cookies.md:
```diff
@@ -52,6 +56,20 @@ Cookie::force_expiry($name, $path = null, $domain = null);
 
+### Samesite attribute
+
+The `samesite` attribute is set on all cookies with a default value of `Lax`. You can change the default value by setting the `default_samesite` value on the
```
```suggestion
The `samesite` attribute is set on all cookies (except the Session cookie) with a default value of `Lax`. You can change the default value by setting the `default_samesite` value on the
```

GuySartorelli replied:
The default for session cookies _is_ `Lax` though.

emteknetnz replied:
Sorry, I phrased this wrong.  This paragraph should be something along the lines of changing `default_samesite` will change the default for all cookies except for the Session cookie, because it has separate config

GuySartorelli replied:
I can add it to that paragraph if you think it's needed, but I think that's covered by the `[info]` block below this
> Note that this _doesn't_ apply for the session cookie, which is handled separately. See [Sessions](/developer_guides/cookies_and_sessions/sessions#samesite-attribute).

emteknetnz replied:
ok good enough

emteknetnz reviewed (changes requested):
Looks like the HttpOnly default of the session cookie is being changed with this PR - I made no changes to the config.  Tested in firefox

Without PR
![image](https://user-images.githubusercontent.com/4809037/171063847-45867fe5-9287-4e48-8baa-bff42eeaaac9.png)

With PR
![image](https://user-images.githubusercontent.com/4809037/171063857-cf376448-df74-4511-87ad-5dbd5040a6fc.png)

emteknetnz commented on src/Control/Session.php:
```diff
@@ -354,6 +368,17 @@ public function start(HTTPRequest $request)
+    private function isCookieSecure(string $sameSite, bool $isHttps): bool
+    {
+        if ($sameSite === 'None') {
+            return true;
```
What happens when $isHttps is false?

GuySartorelli replied:
According to the spec, `secure` _must_ be true when `samesite` is "None". It doesn't specify how or whether this interacts in some way with the `httponly` attribute.
In short: I don't know, but I don't think it's relevant to this return value.

emteknetnz replied:
That makes sense, you'll need to update docs + unit tests accordingly

GuySartorelli replied:
This change requires keeping the private `getSameSite()` method in the next major, so it can't be deprecated anymore. I've created #10342 to capture the intention for adding a `$sameSite` parameter in the next major.

GuySartorelli replied:
We've talked a bit offline about how to handle this situation. There's a chance that `Director::is_https()` incorrectly returns false where https is terminated at a load balancer or other proxy. Because of this, we don't want to throw exceptions - that would be causing errors for actually valid situations. Instead, we want to log a warning (in case it is actually incorrect configuration) but let the cookie be set.
The browser then will reject the cookie (with a warning in the console) if it is indeed invalid, but will accept the cookie if it is valid.

emteknetnz commented on src/Control/Session.php:
```diff
@@ -307,7 +312,16 @@ public function start(HTTPRequest $request)
             if (!headers_sent()) {
-                session_set_cookie_params($timeout ?: 0, $path, $domain ?: null, $secure, true);
+                $sameSite = static::config()->get('cookie_samesite');
+                $secure = $this->isCookieSecure($sameSite, Director::is_https($request));
```
Will `Director::is_https($request)` handle the scenario where https is terminated at a load balancer?

GuySartorelli replied:
I have no idea. But changing that behaviour is out of scope for this PR, I'd say.

emteknetnz commented on src/Control/CookieJar.php:
```diff
@@ -180,4 +187,17 @@ protected function outputCookie(
+     *
+     * @deprecated 5.0 The relevant methods will include a `$sameSite` parameter instead.
+     */
+    private function getSameSite(string $name): string
```
This private function can be unit tested by using using reflection to change its visibility

GuySartorelli replied:
Test added

emteknetnz commented on src/Control/Session.php:
```diff
@@ -354,6 +368,17 @@ public function start(HTTPRequest $request)
+    /**
+     * Determines what the value for the `secure` cookie attribute should be.
+     */
+    private function isCookieSecure(string $sameSite, bool $isHttps): bool
```
This private function can be unit tested by using using reflection to change its visibility

GuySartorelli replied:
Test added

emteknetnz commented on src/Control/Session.php:
```diff
@@ -307,7 +312,16 @@ public function start(HTTPRequest $request)
+                $sameSite = static::config()->get('cookie_samesite');
+                $secure = $this->isCookieSecure($sameSite, Director::is_https($request));
+                session_set_cookie_params([
+                    'lifetime' => $timeout ?: 0,
```
We can add unit testing for this (and CookieJar::outputCookie) by creating private functions to generate the arrays to pass to session_set_cookie_params/setcookie and then using reflection to set their visibility to public

GuySartorelli replied:
Done. For `CookieJar::outputCookie()` the only piece of actual functionality there was whether secure is true or false, so I've extracted only that part into its own private method.

emteknetnz commented on tests/php/Control/CookieTest.php:
```diff
@@ -201,4 +206,57 @@ public function testForceExpiry()
+        Injector::inst()->registerService($mockLogger, LoggerInterface::class);
+
+        // Only samesite === 'None' should log a warning on insecure requests
+        Director::config()->set('alternate_base_url', 'http://insecure.example.com/');
```
Split this one off to a new test method and use 
`$this->expectException(MyException::class);`
`$this->expectExceptionMessage($someMessage);`

GuySartorelli replied:
I'll do the same for the similar session test.

emteknetnz commented on tests/php/Control/CookieTest.php:
```diff
@@ -201,4 +206,57 @@ public function testForceExpiry()
+
+        // There should be no warnings logged for secure requests
+        Director::config()->set('alternate_base_url', 'https://secure.example.com/');
+        Cookie::validateSameSite('None');
```
This is pretty weird.  I'm thinking update Cookie::validateSameSite() to return a bool so it's more testable and behaves like people normally expect a valiate*() function to operate, even though we don't actually use the validate return value

GuySartorelli replied:
I don't think we should return a value just because other similarly named methods return values.... and it wouldn't make it more testable so much as it would lend itself to writing incorrect tests (testing a return value that is never used is a great way to have tests passing when functionality is broken).
Maybe it should be renamed instead? Though I can't think of another name that makes more sense.

GuySartorelli commented on src/Control/Cookie.php:
```diff
@@ -92,4 +100,37 @@ public static function force_expiry($name, $path = null, $domain = null, $secure
+     */
+    public static function validateSameSite(string $sameSite): void
+    {
+        $validValues = [
+            'Strict',
+            'Lax',
+            'None',
+        ];
+        if (!in_array($sameSite, $validValues)) {
```
Is it faster enough to be worth making it marginally less readable? I look at this and go "Why are those set to true? Those values aren't used." We're using `in_array()` in probably lots of places that _could_ be implemented this way if we're that worried about the performance implication but I really don't think it makes enough of an impact to matter?

emteknetnz replied:
Absolutely minimal difference, I'm not too fussed. I usually use the assoc_array method myself that's all

GuySartorelli commented on src/Control/Cookie.php:
```diff
@@ -92,4 +100,37 @@ public static function force_expiry($name, $path = null, $domain = null, $secure
+        if (Controller::has_curr()) {
+            $request = Controller::curr()->getRequest();
+        }
+        return ($request instanceof NullHTTPRequest) ? null : $request;
```
A `NullHTTPRequest` will _always_ have its `scheme` set to "http". This means Director won't fall back to checking `default_base_url` because it expects a real request to be passed in with the scheme used for that request. I don't know if that's a bug in `Director` or not.... I don't know what implications that would have for anything else if we did change it there.

GuySartorelli replied:
I've added a comment above this line explaining why we do this.

emteknetnz reviewed (approved):
Test locally, works great. Thanks for implementing all the feedback!

emteknetnz reviewed (changes requested):
Just noticed there's a failing unit test - happy to merge once this is fixed

GuySartorelli commented:
It turns out the `$this->once()` counts as an assertion (that the method will be called once).
Changed to `$this->never()` and removed the `@doesNotPerformAssertions` annotation.
